R/write_output.R

```diff
@@ -0,0 +1,304 @@
+#'
+#' @return A data.table with merged posterior draws and standardized parameter
+#' names.
+#' @noRd
+post_process_and_merge <- function(draws, fact_table) {
+  # Step 1: Left join the date-time-parameter map onto the Stan draws
+  merged_dt <- merge(
+    draws,
+    fact_table,
+    by = c("time", ".variable"),
+    all.x = TRUE,
+    all.y = FALSE
+  )
+
+  # Step 2: Standardize parameter names
+  data.table::set(merged_dt, j = ".variable", value = factor(
+    merged_dt[[".variable"]],
+    levels = c("imputed_reports", "obs_reports", "R", "r"),
+    labels = c("latent_cases", "obs_cases", "Rt", "growth_rate")
+  ))
+
+  # Step 3: Rename columns as necessary
+  data.table::setnames(
+    merged_dt,
+    old = c(
+      ".draw", ".chain", ".variable", ".value", ".lower", ".upper", ".width",
+      ".point", ".interval", "date", ".iteration"
+    ),
+    new = c(
+      "_draw", "_chain", "_variable", "_value", "_lower", "_upper", "_width",
+      "_point", "_interval", "reference_date", "_iteration"
+    ),
+    # If using summaries, skip draws-specific names
+    skip_absent = TRUE
+  )
+
+  return(merged_dt)
+}
+
+#' Process posterior samples from a Stan fit object (raw draws).
+#'
+#' Extracts raw posterior samples from a Stan fit object and post-processes
+#' them, including merging with a fact table and standardizing the parameter
+#' names. If calling `process_quantiles()` the 50% and 95% intervals are
+#' returned in `{tidybayes}` format.
+#'
+#' @param fit An EpiNow2 fit object with posterior estimates.
+#'
+#' @return A data.table of posterior draws or quantiles, merged and processed.
+#'
+#' @name sample_processing_functions
+NULL
+
+#' @rdname sample_processing_functions
+#' @export
+process_samples <- function(fit) {
+  draws_list <- extract_draws_from_fit(fit)
+  raw_processed_output <- post_process_and_merge(
+    draws_list$stan_draws, draws_list$fact_table
+  )
+  return(raw_processed_output)
+}
+
+#' @rdname sample_processing_functions
+#' @export
+process_quantiles <- function(fit) {
+  # Step 1: Extract the draws
+  draws_list <- extract_draws_from_fit(fit)
+
+  # Step 2: Summarize the draws
+  .variable <- time <- NULL # nolint
+  summarized_draws <- draws_list$stan_draws |>
+    dplyr::group_by(.variable, time) |>
+    tidybayes::median_qi(
+      .width = c(0.5, 0.95),
+    ) |>
+    data.table::as.data.table()
+
+  # Step 3: Post-process summarized draws
+  post_process_and_merge(summarized_draws, draws_list$fact_table)
+}
+
+write_parquet <- function(data, path) {
+  # This is bad practice but `dbBind()` doesn't allow us to parameterize COPY
+  # ... TO.  The danger of doing it this way seems quite low risk because it's
+  # ephemeral from a temporary in-memory DB. There's no actual database to
+  # guard against a SQL injection attack and all the data are already available
+  # here.
+  query <- paste0("COPY (SELECT * FROM df) TO '", path, "' (FORMAT PARQUET)")
```

Comment:
```suggestion
  query <- paste0("COPY (SELECT * FROM df) TO '", path, "' (FORMAT PARQUET, CODEC 'zstd')")
```

Not sure how important this is to us, but I get consistently smaller file sizes when I tell duckdb to use `zstd` encoding.